reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java

```diff
@@ -160,61 +146,159 @@ BoundedScheduledExecutorService createBoundedExecutorService() {
-						} // else swallow
-					}
-				}
-				else {
-					e.shutdownNow();
+
+			if (STATE.compareAndSet(this, a, b)) {
+				try {
+					b.currentResource.boundedServices.evictor.scheduleAtFixedRate(
+							b.currentResource.boundedServices::eviction,
+							ttlMillis, ttlMillis, TimeUnit.MILLISECONDS
+					);
+					return;
+				} catch (RejectedExecutionException ree) {
+					// The executor was most likely shut down in parallel.
+					// If the state is SHUTDOWN/SHUTTING_DOWN - it's ok, no eviction schedule required;
+					// If it's running - the other thread did a restart and will run its' own schedule.
+					// In both cases it's safe to return.
+					return;
 				}
+			}
+
+			if (a.currentResource.boundedServices != SHUTDOWN) {
+				b.currentResource.boundedServices.evictor.shutdownNow();
 				return;
 			}
 		}
 	}
 
+	@Override
+	public boolean await(BoundedServicesState resource, long timeout, TimeUnit timeUnit)
+			throws InterruptedException {
+		if (!resource.boundedServices.evictor.awaitTermination(timeout, timeUnit)) {
+			return false;
+		}
+		if (resource.disposedBoundedStates != null) {
+			for (BoundedState bs : resource.disposedBoundedStates) {
+				if (!bs.executor.awaitTermination(timeout, timeUnit)) {
+					return false;
+				}
+			}
+		}
+		return true;
+	}
+
 	@Override
 	public void dispose() {
-		BoundedServices services = BOUNDED_SERVICES.get(this);
-		if (services != SHUTDOWN && BOUNDED_SERVICES.compareAndSet(this, services, SHUTDOWN)) {
-			ScheduledExecutorService e = EVICTOR.getAndSet(this, null);
-			if (e != null) {
-				e.shutdownNow();
+		for (;;) {
+			SchedulerState<BoundedServicesState> previous = state;
+
+			if (previous.currentResource.boundedServices == SHUTTING_DOWN) {
+				continue;
+			}
+
+			if (previous.currentResource.boundedServices == SHUTDOWN ) {
+				// A dispose process might be ongoing, but we want a forceful shutdown,
+				// so we do our best to release the resources without updating the state.
+				if (previous.initialResource != null) {
+					previous.initialResource.boundedServices.evictor.shutdownNow();
+					if (previous.initialResource.disposedBoundedStates != null) {
+						for (BoundedState bs : previous.initialResource.disposedBoundedStates) {
+							bs.shutdown(true);
+						}
+					}
+				}
+				return;
+			}
+
+			SchedulerState<BoundedServicesState> shuttingDown = SchedulerState.transition(
+					previous.currentResource, SHUTTING_DOWN_STATE, this
+			);
+			if (STATE.compareAndSet(this, previous, shuttingDown)) {
+				assert shuttingDown.initialResource != null;
+				final BoundedState[] toAwait =
+						shuttingDown.initialResource.boundedServices.dispose(true);
+				shuttingDown.initialResource.boundedServices.evictor.shutdownNow();
+
+				SchedulerState<BoundedServicesState> shutdown = SchedulerState.transition(
+						disposed(shuttingDown.initialResource.boundedServices, toAwait),
+						SHUTDOWN_STATE,
+						this
+				);
+				STATE.lazySet(this, shutdown);
```

Comment:
The lazySet is misused. Usually, it is combined with another strong set barrier, ensuring lazySet is flushed out. The nature of the lazySet barrier allows delaying this volatility, thus a thread that is spinning inside an infinite loop may burn a lot of cycles. The best here would be the usage of compareAndSet to ensure the operation is not overwriting state C which happened between your setA and setB

Reply:
Thanks for explaining. To my understanding there would not be a C in the current implementation, as the `SHUTTING_DOWN_STATE` is guarding any thread from injecting a C value before the lazy or eager set has effect. With the suggestion you made in `start()` it would now be possible, so for sure this logic needs to change. Further discussion follows in the suggested code snippet at the end of `disposeGracefully()`.